content/blog/2025-12-16-facade-layout-compiler.md

```diff
@@ -0,0 +1,409 @@
+- **RowDecl:** A row with its number and list of cells
+- **Cell:** A single symbol at a source location
+- **RepeatExpr:** A symbol with a repeat count
+
+**Grammar (EBNF):**
+
+```
+program     ::= rule_decl* row_decl+
+rule_decl   ::= 'rule' IDENTIFIER ':' rule_value+
+row_decl    ::= 'row' INTEGER ':' cell+
+cell        ::= SYMBOL | repeat_expr | symbol_sequence
+repeat_expr ::= SYMBOL '*' INTEGER
+```
+
+**Note on Grammar Classification:** This grammar is regular, and the structure is purely sequential without nesting. The language could equivalently be expressed as a regular expression:
+
+```
+(rule IDENTIFIER : value+)* (row INTEGER : (SYMBOL | SYMBOL * INTEGER)+)+
+```
+
+**Note on Parser Classification:** The parser follows a top-down, predictive parsing strategy with one function per major grammar production (`parse()`, `parse_rule()`, `parse_row()`, `parse_cells()`). It is not strictly a recursive-descent parser because no function calls itself—the grammar's repetition operators (`*` and `+`) are implemented with iterative `while` loops rather than recursive calls. This iterative approach is natural for a regular grammar where recursion is unnecessary. The parser uses single-token lookahead via `check()` to predict which production to apply, making it a predictive parser.
+
+**Key design decisions:**
+
+- **Error recovery:** When a syntax error is detected (e.g., missing colon), the parser calls `sync_to_row()` to skip tokens until the next `row` keyword. This prevents cascading errors and allows multiple independent errors to be reported.
+
+- **Repeat expression handling:** The `*0` case is handled by simply not adding any cell to the AST—the absence of cells is the correct representation of "repeat zero times."
+
+- **Duplicate row handling:** When the same row number appears twice, the parser accepts both, only the latest duplicated row would show, and the semantic analyzer emits a warning. This is more user-friendly than treating duplicates as errors.
+
+- **Strict validation:** Row numbers must be positive integers. Syntax errors like `row -1:`, `row 1.5:`, or `row :` are caught and reported with helpful hints.
+
+---
+
+## Middle-End: Semantic Analysis and Normalization
+
+The semantic analyzer transforms the parsed AST into a normalized grid representation and enforces architectural constraints.
+
+### Grid Construction and Normalization
+
+**Step 1: Build initial grid**
+
+The analyzer iterates over row declarations (sorted by row number) and populates a 2D array. If row numbers have gaps (e.g., rows 1, 3, 5 are specified), intermediate rows are initialized as empty lists.
+
+**Step 2: Normalize grid dimensions**
+
+After construction, the grid may be jagged (rows of different lengths). The normalizer:
+
+- Computes the maximum row width
+- Pads shorter rows with `E` (empty) cells
+- Records which cells were auto-filled in metadata
+
+This guarantees a rectangular grid for all downstream processing.
```

Comment:
This seems to more or less repeat what you've said above about how padding works. Is it possible to say all this in 1 or 2 sentences?